### PR TITLE
[Access] Use local event for AccessAPI get events endpoints

### PR DESCRIPTION
**Pull request by peterargue**
Closes: https://github.com/onflow/flow-go/issues/4751

Use locally indexed events for `GetEvents*` endpoints

### CODE REVIEW

File: engine/access/rpc/backend/backend_events.go [hunk 1]
```diff
@@ -106,10 +118,95 @@ func (b *backendEvents) GetEventsForBlockIDs(
...
+	missing := make([]*flow.Header, 0)
+	resp := make([]flow.BlockEvents, 0)
+	for _, header := range blockHeaders {
+		if ctx.Err() != nil {
+			return nil, nil, rpc.ConvertError(ctx.Err(), "failed to get events", codes.Canceled)
+		}
+
+		events, err := b.events.ByBlockID(header.ID())
+		if err != nil {
+			if errors.Is(err, storage.ErrNotFound) {
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_events.go:172]: I'm 90% sure `ErrNotFound` is only returned if events for the block have not been indexed. double check and add a comment explaining

File: engine/access/rpc/backend/backend_events.go [hunk 2]
```diff
@@ -109,10 +121,103 @@ func (b *backendEvents) GetEventsForBlockIDs(
...
+	ctx context.Context,
+	blockHeaders []*flow.Header,
+	eventType string,
+	requiredEventEncodingVersion entities.EventEncodingVersion,
+) ([]flow.BlockEvents, []*flow.Header, error) {
+	target := flow.EventType(eventType)
+
+	missing := make([]*flow.Header, 0)
+	resp := make([]flow.BlockEvents, 0)
+	for _, header := range blockHeaders {
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/access/rpc/backend/backend_events.go:175]: the caller checks that

here
https://github.com/onflow/flow-go/blob/ce9c12a84e4697134806bfe8797dd1377b8907b5/engine/access/rpc/backend/backend_events.go#L52-L55

and here
https://github.com/onflow/flow-go/blob/ce9c12a84e4697134806bfe8797dd1377b8907b5/engine/access/rpc/backend/backend_events.go#L109-L111

File: engine/access/rpc/backend/backend_events.go [hunk 3]
```diff
@@ -109,10 +121,103 @@ func (b *backendEvents) GetEventsForBlockIDs(
...
+
+// getBlockEventsFromStorage retrieves events for all the specified blocks that have the given type
+// from the local storage
+func (b *backendEvents) getBlockEventsFromStorage(
+	ctx context.Context,
+	blockHeaders []*flow.Header,
+	eventType string,
+	requiredEventEncodingVersion entities.EventEncodingVersion,
+) ([]flow.BlockEvents, []*flow.Header, error) {
+	target := flow.EventType(eventType)
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/access/rpc/backend/backend_events.go:171]: the `target` is only used for effectively a string comparison, so it's pretty benign, but I can add it.

File: cmd/access/node_builder/access_node_builder.go [hunk 4]
```diff
@@ -1060,6 +1061,11 @@ func (builder *FlowAccessNodeBuilder) extraFlags() {
 		flags.StringVar(&builder.registersDBPath, "execution-state-dir", defaultConfig.registersDBPath, "directory to use for execution-state database")
 		flags.StringVar(&builder.checkpointFile, "execution-state-checkpoint", defaultConfig.checkpointFile, "execution-state checkpoint file")
 
+		flags.StringVar(&builder.rpcConf.BackendConfig.EventQueryMode,
+			"event-query-mode",
+			defaultConfig.rpcConf.BackendConfig.EventQueryMode,
+			"mode to use when querying events. one of (local-only, execution-nodes-only, failover)")
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:1067]:
```suggestion
			"mode to use when querying events. one of [local-only, execution-nodes-only(default), failover]")
```

**Review by zhangchiqing — APPROVED** [review]

**Pull request merged**